base/buffer.3
.\" vim: tw=75
.TH buffer 3
.SH NAME
buffer \- generic read/write buffering
.SH SYNTAX
.B #include \(dqbuffer.h\(dq

buffer* buffer_0;   /* like stdio's stdin */
.br
buffer* buffer_1;   /* like stdio's stdout */
.br
buffer* buffer_2;   /* like stdio's stderr */

void \fBbuffer_init\fR(buffer &\fIb\fR,ssize_t (*\fIop\fR)(int,char *,size_t),
                int \fIfd\fR, char *\fIy\fR, size_t \fIylen\fR);
.br
ssize_t \fBbuffer_get\fP(buffer *\fIb\fR,char *\fIx\fR,size_t \fIlen\fR);

int \fBbuffer_put\fP(buffer *\fIb\fR,const char *\fIx\fR,size_t \fIlen\fR);
.br
int \fBbuffer_puts\fP(buffer *\fIb\fR,const char *\fIx\fR);
.br
int \fBbuffer_putalign\fP(buffer *\fIb\fR,char *\fIx\fR,unsigned int \fIlen\fR);
.br
int \fBbuffer_putsalign\fP(buffer *\fIb\fR,char *\fIx\fR);

int \fBbuffer_putflush\fP(buffer *\fIb\fR,char *\fIx\fR,unsigned int \fIlen\fR);
.br
int \fBbuffer_putsflush\fP(buffer *\fIb\fR,char *\fIx\fR);

int \fBbuffer_flush\fP(buffer *\fIb\fR);
.br
int \fBbuffer_copy\fP(buffer *\fIbo\fR,buffer *\fIbi\fR);

int \fBbuffer_unixread\fP(int \fIfd\fR,char *\fIbuf\fR,size_t \fIlen\fR);
.br
int \fBbuffer_unixwrite\fP(int \fIfd\fR,char *\fIbuf\fR,size_t \fIlen\fR);

.SH DESCRIPTION
\fBbuffer.h \fR describes a generic buffer interface that can be used for
read and write buffering. Buffers must be initialized with
\fBbuffer_init\fR.

A buffer can only be used for reading or writing at the same time, not
both.

Unlike \fBstdio\fR, these write buffers are not flushed automatically at
program termination; you must manually call \fBbuffer_flush\fR,
\fBbuffer_putflush\fR, or \fBbuffer_putsflush\fR.

\fBbuffer_init\fR prepares \fIb\fR to store a string in \fIy\fR[0],
\fIy\fR[1], ..., \fIy\fR[\fIylen\fR-1]. Initially the string is empty.

\fBbuffer_init\fR also prepares \fIb\fR to use the read/write operation
specified by \fIop\fR and \fIfd\fR.

You can use

  buffer \fIb\fR = BUFFER_INIT(\fIop\fR,\fIfd\fR,\fIy\fR,\fIylen\fR);

to initialize \fIb\fR statically if \fIop\fR, \fIfd\fR, \fIy\fR, and \fIylen\fR
are compile-time constants.

You can call \fBbuffer_init\fR again at any time. Note that this discards
the currently buffered string.

\fBbuffer_get\fR copies data to \fIx\fR[0], \fIx\fR[1], ...,
\fIx\fR[\fIlen\fR-1] from the beginning of a string stored in preallocated
space; removes these \fIlen\fR bytes from the string; and returns
\fIlen\fR.

If, however, the string has fewer than \fIlen\fR (but more than 0) bytes,
\fIbuffer_get\fR copies only that many bytes, and returns that number.

If the string is empty, \fBbuffer_get\fR first uses a \fBread operation\fR
to feed data into the string. The \fBread operation\fR may indicate end of
input.

The preallocated space and the \fBread operation\fR are specified by
\fIb\fR. You must initialize \fBb\fR using \fBbuffer_init\fR before
calling \fBbuffer_get\fR (or use the pre-initialized \fIbuffer_0\fR).

\fBbuffer_put\fR writes \fIlen\fR bytes from \fIx\fR to \fIb\fR.

The difference to \fBbuffer_putalign\fR is that, when there isn't enough
space for new data, \fBbuffer_put\fR calls \fBbuffer_flush\fR before
copying any data, while \fBbuffer_putalign\fR fills all available space
with data before calling \fBbuffer_flush\fR

\fBbuffer_copy\fR copies one buffer to other one.  The output buffer needs
to have at least the preallocated size of the input buffer.
\fBbuffer_unixread\fR and \fBbuffer_unixwrite\fR perform the same operation
like standard Unix \fBread\fR or \fBwrite\fR.

.SH MACROS
Apart from this basic usage, some helpful macro definitions are provided:
\fBBUFFER_INIT(op,fd,buf,len)\fR uses \fIop\fR function to operate
(read/write) on \fIbuf\fR with \fIlen\fR to file descriptor \fIfd\fR.
\fBbuffer_GETC(buf,c)\fR returns the upmost position of character \fIc\fR
within buffer \fIbuf\fR or 0. \fBbuffer_PUTC(buf,c)\fR adds character
\fIc\fR to buffer \fIbuf\fR.

.SH EXAMPLE
  #include <buffer.h>
  #include <open.h>

  char buf[BUFFER_INSIZE]; // defined in buffer.h
  int fd = open_read("/etc/services");
  buffer input;

  if (fd >= 0) {
    char x;
    buffer_init(&input,read,fd,buf,sizeof buf);
    while (buffer_get(&input,&x,1) == 1) {
      buffer_put(buffer_1,&x,1);
      if (x == '\\n') break;
    }
    buffer_flush(buffer_1);
  }

.SH "RETURN CODES"
\fBbuffer_put\fR and \fBbuffer_get\fR return \fI0\fR if everything was
fine, \fI-1\fR, on error (setting \fIerrno\fR). \fBbuffer_copy\fR returns
\fI-2\fR, if the input buffer can't be read, and \fI-3\fR, if the data
can't successfully copied to the output buffer. On success
\fBbuffer_copy\fR returns \fI0\fR.

.SH "REFERENCES"
https://cr.yp.to/lib/buffer_get.html

https://cr.yp.to/lib/buffer_put.html
.SH "SEE ALSO"
stdio(3)


ChangeLog
* ChangeLog
Release 1.0 Start 21/05/2020
01. added --release option to specify release on command line
- 13/09/2020
02. libqmail.spec: added Provides: qmail-libs
- 15/09/2020
03. added mess822, commands, rewritehost, sconfig functions from indimail-mta
- 16/09/2020
04. FreeBSD fixes
- 01/10/2020
05. Darwin Port (fmt_ushort.c, default.configure, configure.ac)
- 04/10/2020
06. fixed broken crypt(3) implemetation on Darwin Mac OS X
07. reorgnized directory structure. All qmail functions moved to 'base'
08. textcode/fmt_utf8.c: updated with upstream libowfat-0.32
- 09/10/2020
09. replaced fmt_tohex() with tohex()
10. global.h, hmac_md5.c, hmac_ripemd.c, md5.c, md5.h: replaced updated
    functions from indimail-mta
- 18/10/2020
11. alloc_re.c: fix overwriting unallocate memory location when
    new_size < old_size


base/buffer.c
/*
 * $Id: buffer.c,v 1.1 2024-01-22 10:15:46+05:30 Cprogrammer Exp mbhangui $
 */
#include <unistd.h>
#include <errno.h>
#include "buffer.h"
#include <str.h>
#include <byte.h>

/*
 * author djb
 */

void
buffer_init(buffer *s, ssize_t(*op) (), int fd, char *buf, size_t len)
{
	s->x = buf;
	s->fd = fd;
	s->op = op;
	s->p = 0;
	s->n = len;
}

ssize_t
buffer_0_read(int fd, char *buf, size_t len)
{
	if (buffer_flush(buffer_1) == -1)
		return -1;
	return read(fd, buf, len);
}

char            buffer_0_space[BUFFER_INSIZE];
char            buffer_1_space[BUFFER_OUTSIZE];
char            buffer_2_space[BUFFER_OUTSIZE];
static buffer   it0 = BUFFER_INIT(buffer_0_read, 0, buffer_0_space, sizeof (buffer_0_space));
static buffer   it1 = BUFFER_INIT(write, 1, buffer_1_space, sizeof (buffer_1_space));
static buffer   it2 = BUFFER_INIT(write, 2, buffer_2_space, sizeof (buffer_2_space));

buffer         *buffer_0 = &it0;
buffer         *buffer_1 = &it1;
buffer         *buffer_2 = &it2;

char            buffer_0_small[BUFFER_SMALL];
char            buffer_1_small[BUFFER_SMALL];
char            buffer_2_small[BUFFER_SMALL];

static buffer   is0 = BUFFER_INIT(buffer_0_read, 0, buffer_0_small, sizeof (buffer_0_small));
static buffer   is1 = BUFFER_INIT(write, 1, buffer_1_small, sizeof (buffer_1_small));
static buffer   is2 = BUFFER_INIT(write, 2, buffer_2_small, sizeof (buffer_2_small));

buffer         *buffer_0small = &is0;
buffer         *buffer_1small = &is1;
buffer         *buffer_2small = &is2;

ssize_t
buffer_unixread(int fd, char *buf, size_t len)
{
	return read(fd, buf, len);
}

ssize_t
buffer_unixwrite(int fd, char *buf, size_t len)
{
	return write(fd, buf, len);
}

int
buffer_copy(buffer *bout, buffer *bin)
{
	int             n;
	char           *x;

	for (;;) {
		if ((n = buffer_feed(bin)) < 0)
			return -2;
		if (!n)
			return 0;
		x = buffer_PEEK(bin);
		if (buffer_put(bout, x, n) == -1)
			return -3;
		buffer_SEEK(bin, n);
	}
}

static int
oneread(ssize_t(*op) (), int fd, char *buf, size_t len)
{
	int             r;

	for (;;) {
		if ((r = op(fd, buf, len)) == -1) {
			if (errno == EINTR)
				continue;
		}
		return r;
	}
}

static int
getthis(buffer *s, char *buf, size_t len)
{
	if (len > s->p)
		len = s->p;
	s->p -= len;
	byte_copy(buf, len, s->x + s->n);
	s->n += len;
	return len;
}

int
buffer_feed(buffer *s)
{
	int             r;

	if (s->p)
		return s->p;
	if ((r = oneread(s->op, s->fd, s->x, s->n)) <= 0)
		return r;
	s->p = r;
	s->n -= r;
	if (s->n > 0)
		byte_copyr(s->x + s->n, r, s->x);
	return r;
}

int
buffer_bget(buffer *s, char *buf, size_t len)
{
	int             r;

	if (s->p > 0)
		return getthis(s, buf, len);
	if (s->n <= len)
		return oneread(s->op, s->fd, buf, s->n);
	if ((r = buffer_feed(s)) <= 0)
		return r;
	return getthis(s, buf, len);
}

int
buffer_get(buffer *s, char *buf, size_t len)
{
	int             r;

	if (s->p > 0)
		return getthis(s, buf, len);
	if (s->n <= len)
		return oneread(s->op, s->fd, buf, len);
	if ((r = buffer_feed(s)) <= 0)
		return r;
	return getthis(s, buf, len);
}

char           *
buffer_peek(buffer *s)
{
	return s->x + s->n;
}

void
buffer_seek(buffer *s, size_t len)
{
	s->n += len;
	s->p -= len;
}

static int
allwrite(ssize_t(*op) (), int fd, const char *buf, size_t len)
{
	int             w;

	while (len) {
		if ((w = op(fd, buf, len)) == -1) {
			if (errno == EINTR)
				continue;
			return -1; /* note that some data may have been written */
		}
		if (w == 0) /* luser's fault */
			;
		buf += w;
		len -= w;
	}
	return 0;
}

void
buffer_discard(buffer *s)
{
	s->p = 0;
	return;
}

int
buffer_flush(buffer *s)
{
	int             p;

	if (!(p = s->p))
		return 0;
	s->p = 0;
	return allwrite(s->op, s->fd, s->x, p);
}

int
buffer_putalign(buffer *s, const char *buf, size_t len)
{
	unsigned int    n;

	while (len > (n = s->n - s->p)) {
		byte_copy(s->x + s->p, n, buf);
		s->p += n;
		buf += n;
		len -= n;
		if (buffer_flush(s) == -1)
			return -1;
	}
	/*- now len <= s->n - s->p */
	byte_copy(s->x + s->p, len, buf);
	s->p += len;
	return 0;
}

int
buffer_put(buffer *s, const char *buf, size_t len)
{
	unsigned int    n;

	n = s->n;
	if (len > n - s->p) {
		if (buffer_flush(s) == -1)
			return -1;
		/*- now s->p == 0 */
		if (n < BUFFER_OUTSIZE)
			n = BUFFER_OUTSIZE;
		while (len > s->n) {
			if (n > len)
				n = len;
			if (allwrite(s->op, s->fd, buf, n) == -1)
				return -1;
			buf += n;
			len -= n;
		}
	}
	/*- now len <= s->n - s->p */
	byte_copy(s->x + s->p, len, buf);
	s->p += len;
	return 0;
}

int
buffer_putflush(buffer *s, const char *buf, size_t len)
{
	if (buffer_flush(s) == -1)
		return -1;
	return allwrite(s->op, s->fd, buf, len);
}

int
buffer_putsalign(buffer *s, const char *buf)
{
	return buffer_putalign(s, buf, str_len(buf));
}

int
buffer_puts(buffer *s, const char *buf)
{
	return buffer_put(s, buf, str_len(buf));
}

int
buffer_putsflush(buffer *s, const char *buf)
{
	return buffer_putflush(s, buf, str_len(buf));
}

/*
 * $Log: buffer.c,v $
 * Revision 1.1  2024-01-22 10:15:46+05:30  Cprogrammer
 * Initial revision
 *
 */
